core-bundle/contao/drivers/DC_Table.php

```diff
@@ -2209,62 +2195,52 @@ public function edit($intId=null, $ajaxId=null)
 		// Versions overview
 		if (($GLOBALS['TL_DCA'][$this->strTable]['config']['enableVersioning'] ?? null) && !($GLOBALS['TL_DCA'][$this->strTable]['config']['hideVersionMenu'] ?? null))
 		{
-			$version = $objVersions->renderDropdown();
+			$parameters['version_dropdown'] = $objVersions->renderDropdown();
 		}
 		else
 		{
-			$version = '';
+			$parameters['version_dropdown'] = '';
 		}
 
+		// Form settings and buttons
 		$security = System::getContainer()->get('security.helper');
 
-		$strButtons = System::getContainer()
-			->get('contao.data_container.buttons_builder')
-			->generateEditButtons(
-				$this->strTable,
-				(bool) $this->ptable,
-				$security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, $this->addDynamicPtable(array('pid' => $this->intCurrentPid)))),
-				$security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, array_replace($currentRecord, array('id' => null, 'sorting' => null)))),
-				$this
-			);
-
-		// Add the buttons and end the form
-		$return .= '
-</div>
-  ' . $strButtons . '
-</form>';
-
-		$strVersionField = '';
-
-		// Store the current version number (see #8412)
-		if ($intLatestVersion !== null)
-		{
-			$strVersionField = '
-<input type="hidden" name="VERSION_NUMBER" value="' . $intLatestVersion . '">';
-		}
-
+		$parameters['form'] = array(
+			'multipart' => $this->blnUploadable,
+			'onsubmit' => $this->onsubmit,
+			'buttons' => System::getContainer()
+				->get('contao.data_container.buttons_builder')
+				->generateEditButtons(
+					$this->strTable,
+					(bool) $this->ptable,
+					$security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, $this->addDynamicPtable(array('pid' => $this->intCurrentPid)))),
+					$security->isGranted(ContaoCorePermissions::DC_PREFIX . $this->strTable, new CreateAction($this->strTable, array_replace($currentRecord, array('id' => null, 'sorting' => null)))),
+					$this
+				),
```

Comment:
If you really do want to enforce the `__toString()` (which I do not agree with, and I don't think we do anywhere else in the system – but that's an implementation detail), the it should also be done here.

Reply:
As mentioned in https://github.com/contao/contao/pull/8981#discussion_r2509675994, I did this for two reasons:

1) The order of operations changes, because before there was an implicit string cast due to concatenation which evaluated the code to generate buttons at that time.

2) The types of template variables must not be of different kinds (`string` vs `DataContainerGlobalOperationsBuilder`), because you cannot test for the type. But if 1) is not an issue, we can surely go with `DataContainerGlobalOperationsBuilder|null` instead.

Reply:
Did the latter in 49566d75a77cb16e0f7dcc2cea8b23a0bbeb2469.